util/logtools/make_timeline.ts

```diff
@@ -0,0 +1,475 @@
+  const entry: TimelineEntry = {
+    time: matches.timestamp,
+    combatant: matches.source,
+    abilityId: matches.id,
+    abilityName: abilityName,
+    lineType: 'ability',
+  };
+  return entry;
+};
+
+const extractRawLines = async (fileName: string, start: string, end: string): Promise<string[]> => {
+  const lines: string[] = [];
+  const file = readline.createInterface({
+    input: fs.createReadStream(fileName),
+  });
+  let started = false;
+  for await (const line of file) {
+    // This will fail on lines with 3-digit identifiers,
+    // but that's okay because those will never be start lines.
+    const lineTimeStamp = line.substring(14, 26);
+    if (start === lineTimeStamp && !started)
+      started = start === lineTimeStamp;
+    if (started)
+      lines.push(line);
+    if (lineTimeStamp === end)
+      file.close();
+  }
+  return lines;
+};
+
+const extractTLEntries = async (
+  fileName: string,
+  start: Date,
+  end: Date,
+  targetArray?: string[],
+): Promise<TimelineEntry[]> => {
+  const entries: TimelineEntry[] = [];
+  const fileStart = TLFuncs.timeFromDate(start);
+  const fileEnd = TLFuncs.timeFromDate(end);
+  const lines: string[] = await extractRawLines(fileName, fileStart, fileEnd);
+
+  // We have exactly and only the lines from the start to the end of the encounter.
+  for (const line of lines) {
+    const targetable = NetRegexes.nameToggle().exec(line)?.groups;
+    const ability = NetRegexes.ability().exec(line)?.groups;
+    // Cull non-relevant lines immediately.
+
+    if (!ability && !targetable)
+      continue;
+
+    // Make nameplate toggle lines if and only if the user has specified them.
+    if (targetable) {
+      if (targetArray && targetArray.includes(targetable.name)) {
+        const targetEntry = parseNameToggleToEntry(targetable);
+        entries.push(targetEntry);
+      }
+      continue;
+    }
+
+    // At this point, only ability lines are left.
+    if (ability) {
+      // Cull non-enemy lines
+      // TODO: Handle this using the raid eumlator's line parsing functionality.
+      if (ability.sourceId.substring(0, 3) !== '400')
+        continue;
+      const abilityEntry = parseAbilityToEntry(ability);
+      entries.push(abilityEntry);
+      continue;
+    }
+
+    // There shouldn't be any way that we get here,
+    // but if we do, something is drastically wrong.
+    // Notify the user of the malformed line and continue.
+    const errString = `Warning: Potentially malformed/corrupted log line found:\n${line}\n\n`;
+    console.log(errString);
+    continue;
+  }
+
+  if (entries.length === 0) {
+    console.error('Fight not found');
+    process.exit(-1);
+  } else {
+    return entries;
+  }
+};
+
+const ignoreTimelineAbilityEntry = (entry: TimelineEntry, args: ExtendedArgs): boolean => {
+  const abilityName = entry.abilityName;
+  const abilityId = entry.abilityId;
+  const combatant = entry.combatant;
+
+  const ia = args.ignore_ability as string[] | null;
```

Comment:
This could be a little messy to work around. I think we would need to revisit how we're typing `args` generally, since `argParse` just hands`Any` to us. The best I can think of  is "top-level check that any non-null args are of type X" on a per-script basis. We won't be able to export a validation function like this alongside the basic `TimelineArgs` type since a runtime check isn't going to be good enough.

All that is to say, any such changes we make are going to cascade through a few different files in the `logtools` folder. I'm fine doing that, and it definitely *needs* to be done, but it's going to mean this PR contains a little more than just the addition of `make_timeline`. Alternatively, we can put this one on hold and I can go through and do that cascading cleanup first.

Reply:
Line 123 where you call argparse.parse, that will accept a pre-created namespace object and populate that object with the parameters instead. See:
https://github.com/quisquous/cactbot/blob/c764a17fc1f92e2d73c5b3b0c1f68f9ad613bafa/util/index.ts#L42-L48
https://github.com/quisquous/cactbot/blob/c764a17fc1f92e2d73c5b3b0c1f68f9ad613bafa/util/index.ts#L54
https://github.com/quisquous/cactbot/blob/c764a17fc1f92e2d73c5b3b0c1f68f9ad613bafa/util/index.ts#L87-L90
https://github.com/quisquous/cactbot/blob/c764a17fc1f92e2d73c5b3b0c1f68f9ad613bafa/util/index.ts#L56

This doesn't really directly offer type safety itself, but it at least doesn't require the `... as ...` casts.

Here's a full diff which makes the requisite adjustments. It compiles with `npm run tsc-no-emit`.

```diff
diff --git a/util/logtools/arg_parser.ts b/util/logtools/arg_parser.ts
index 4c932fb21..fc1dd3eb7 100644
--- a/util/logtools/arg_parser.ts
+++ b/util/logtools/arg_parser.ts
@@ -1,14 +1,13 @@
 import argparse from 'argparse';
 
 type TimelineArgs = {
-  [index: string]: string | number | boolean | undefined;
-  'file'?: string;
-  'force'?: boolean;
-  'search_fights'?: number;
-  'search_zones'?: number;
-  'fight_regex'?: string;
-  'zone_regex'?: string;
-  'adjust'?: number;
+  'file': string | null;
+  'force': boolean | null;
+  'search_fights': number | null;
+  'search_zones': number | null;
+  'fight_regex': string | null;
+  'zone_regex': string | null;
+  'adjust': number | null;
 };
 
 class LogUtilArgParse {
diff --git a/util/logtools/make_timeline.ts b/util/logtools/make_timeline.ts
index e1f237c00..bd524ed00 100644
--- a/util/logtools/make_timeline.ts
+++ b/util/logtools/make_timeline.ts
@@ -1,6 +1,8 @@
 import fs from 'fs';
 import readline from 'readline';
 
+import { Namespace } from 'argparse';
+
 import NetRegexes from '../../resources/netregexes';
 import PetData from '../../resources/pet_names';
 import { NetMatches } from '../../types/net_matches';
@@ -37,18 +39,36 @@ type TimelineEntry = {
 };
 
 type ExtendedArgs = TimelineArgs & {
-  [index: string]: string | string[] | number | boolean | undefined;
-  'output-file'?: string;
-  'start'?: string;
-  'end'?: string;
-  'ignore-id'?: string[];
-  'ignore-ability'?: string[];
-  'ignore-combatant'?: string[];
-  'only-combatant'?: string[];
-  'phase'?: string;
-  'include-targetable'?: string[];
+  'output_file': string | null;
+  'start': string | null;
+  'end': string | null;
+  'ignore_id': string[] | null;
+  'ignore_ability': string[] | null;
+  'ignore_combatant': string[] | null;
+  'only_combatant': string[] | null;
+  'phase': string | null;
+  'include_targetable': string[] | null;
 };
 
+class ExtendedArgsNamespace extends Namespace implements ExtendedArgs {
+  'file': string | null;
+  'force': boolean | null;
+  'search_fights': number | null;
+  'search_zones': number | null;
+  'fight_regex': string | null;
+  'zone_regex': string | null;
+  'adjust': number | null;
+  'output_file': string | null;
+  'start': string | null;
+  'end': string | null;
+  'ignore_id': string[] | null;
+  'ignore_ability': string[] | null;
+  'ignore_combatant': string[] | null;
+  'only_combatant': string[] | null;
+  'phase': string | null;
+  'include_targetable': string[] | null;
+}
+
 // Some NPCs can be picked up by our entry processor.
 // We list them out explicitly here so we can ignore them at will.
 const ignoredCombatants = PetData['en'].concat([
@@ -120,7 +140,8 @@ timelineParse.parser.addArgument(['--include_targetable', '-it'], {
   nargs: '+',
   help: 'Set this flag to include "34" log lines when making the timeline',
 });
-const args = timelineParse.parser.parseArgs() as ExtendedArgs;
+const args = new ExtendedArgsNamespace({});
+timelineParse.parser.parseArgs(undefined, args);
 
 const printHelpAndExit = (errString: string): void => {
   console.error(errString);
@@ -134,7 +155,8 @@ if (args.file === null)
 if (!args.file?.includes('.log'))
   printHelpAndExit('Error: Must specify an FFXIV ACT log file, as log.log\n');
 let numExclusiveArgs = 0;
-for (const opt of ['search_fights', 'search_zones']) {
+const exclusiveArgs = ['search_fights', 'search_zones'] as const;
+for (const opt of exclusiveArgs) {
   if (args[opt] !== null)
     numExclusiveArgs++;
 }
@@ -266,10 +288,10 @@ const ignoreTimelineAbilityEntry = (entry: TimelineEntry, args: ExtendedArgs): b
   const abilityId = entry.abilityId;
   const combatant = entry.combatant;
 
-  const ia = args.ignore_ability as string[] | null;
-  const ii = args.ignore_id as string[] | null;
-  const ic = args.ignore_combatant as string[] | null;
-  const oc = args.only_combatant as string[] | null;
+  const ia = args.ignore_ability;
+  const ii = args.ignore_id;
+  const ic = args.ignore_combatant;
+  const oc = args.only_combatant;
   // Ignore auto-attacks named "attack"
   if (abilityName?.toLowerCase() === 'attack')
     return true;
```

Reply:
Well, that's great. Thanks for taking the time to put that together. I wasn't anticipating having much time tonight because of reclears, so that's a lovely savings.